lib/index.ts

```diff
@@ -0,0 +1,554 @@
+			params, callback),
+		update: (params: ConversationRequest.Update, callback?: Callback<void>): Promise<void> =>
+			this.httpCall({ path: 'conversations/<conversation_id}', reqType: 'PATCH' }, params, callback),
+	};
+
+	public inbox = {
+		create: (params: InboxRequest.Create, callback?: Callback<InboxCreation>): Promise<InboxCreation> =>
+			this.httpCall({ path: 'inboxes', reqType: 'POST' }, params, callback),
+		get: (params: InboxRequest.Get, callback?: Callback<Inbox>): Promise<Inbox> =>
+			this.httpCall({ path: 'inboxes/<inbox_id>', reqType: 'GET' }, params, callback),
+		list: (callback?: Callback<Inboxes>): Promise<Inboxes> =>
+			this.httpCall({ path: 'inboxes', reqType: 'GET' }, null, callback),
+		listChannels: (params: InboxRequest.ListChannels, callback?: Callback<InboxChannels>):
+			Promise<InboxChannels> => this.httpCall({ path: 'inboxes/<inbox_id>/channels', reqType: 'GET' },
+			params, callback),
+		listConversations: (params: InboxRequest.ListConversations, callback?: Callback<InboxConversations>):
+			Promise<InboxConversations> => this.httpCall({ path: 'inboxes/<inbox_id>/conversations[q:page:limit]',
+			reqType: 'GET' }, params, callback),
+		listTeammates: (params: InboxRequest.ListTeammates, callback?: Callback<InboxTeammates>):
+			Promise<InboxTeammates> => this.httpCall({ path: 'inboxes/<inbox_id>/teammates', reqType: 'GET' },
+			params, callback),
+	};
+
+	public message = {
+		get: (params: MessageRequest.Get, callback?: Callback<Message>): Promise<Message> =>
+			this.httpCall({ path: 'messages/<message_id>', reqType: 'GET' }, params, callback),
+		receiveCustom: (params: MessageRequest.ReceiveCustom,
+			callback?: Callback<ConversationReference>): Promise<ConversationReference> =>
+			this.httpCall({ path: 'channels/<channel_id>/incoming_messages', reqType: 'POST' }, params, callback),
+		reply: (params: MessageRequest.Reply, callback?: Callback<Status>): Promise<Status> =>
+			this.httpCall({ path: 'conversations/<conversation_id>/messages', reqType: 'POST' }, params, callback),
+		send: (params: MessageRequest.Send, callback?: Callback<ConversationReference>):
+			Promise<ConversationReference> => this.httpCall({ path: 'channels/<channel_id>/messages',
+			reqType: 'POST' }, params, callback),
+	};
+
+	private apiKey: string;
+
+	constructor(apiKey: string) {
+		// Key.
+		this.apiKey = apiKey;
+	}
+
+	private httpCall(details: Request, params: any, callback?: InternalCallback): Promise<any | void>  {
+		const requestOpts = {
+			body: (params) ? params : {},
+			headers: {
+				Authorization: `Bearer ${this.apiKey}`
+			},
+			json: true,
+			method: details.reqType,
+			url: `${URL}/${this.formatPath(details.path, params)}`
+		};
+
+		// Make the request.
+		return request(requestOpts).promise().catch((error: any) => {
+			// Format this into something useful, if we can.
+			const frontError: FrontError = error.error._error;
+			let newError: FrontError = Object.create(Error.prototype);
```

Comment:
If you use [typed-error](https://www.npmjs.com/package/typed-error) then you should be able to use
```typescript
class FrontError extends TypedError {
  ...
}
...
let newError = new FrontError(frontError);
```
(and the extra properties can either be handled in the FrontError constructor or with the same assignment as below - I think in the constructor makes more sense though)

Also this way the newly constructed error will keep the stack trace from the existing one, and `newError instanceof FrontError` will be true - if you use `extends Error` directly that check will fail because of the way typescript `extends` inheritance interacts with the immutable Error object (it's possible that's been changed in a new typescript though)